keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,935 @@
+specific ClusterIP that has already been allocated will return an error.
+
+Current implementation of the Services IP allocation logic has several limitations:
+
+- users can not resize or increase the ranges of IPs assigned to Services, causing problems when
+  there are overlapping networks or the cluster run out of available IPs.
+- the Service IP range is not exposed, so it is not possible for other components in the cluster to
+  consume it
+- the configuration is per node, there is no consensus and different apiservers can fight and delete
+  others IPs.
+- the allocation logic is racy, with the possibility of leaking ClusterIPs
+  https://github.com/kubernetes/kubernetes/issues/87603
+- the size of the Service Cluster CIDR, for IPv4 the prefix size is limited to /12, however, for
+  IPv6 it is limited to /112 or fewer.This restriction is causing issues for IPv6 users, since /64
+  is the standard and minimum recommended prefix length
+- only is possible to use reserved a range of Service IP addresses using the feature gate
+  "keps/sig-network/3070-reserved-service-ip-range"
+  https://github.com/kubernetes/kubernetes/issues/95570
+
+### Goals
+
+Implement a new allocation logic for Services IPs that:
+
+- scales well with the number of reservations
+- the number of allocation is tunable
+- is completely backwards compatible
+
+### Non-Goals
+
+- Any change unrelated to Services
+- Any generalization of the API model that can evolve onto something different, like an IPAM API, or
+  collide with existing APIs like Ingress and GatewayAPI.
+- NodePorts use the same allocation model than ClusterIPs, but it is totally out of scope of this
+  KEP. However, this KEP can be a proof that a similar model will work for NodePorts too.
+- Change the default IPFamily used for Services IP allocation, the defaulting depends on the
+  services.spec.IPFamily and services.spec.IPFamilyPolicy, a simple webhook or an admission plugin
+  can set this fields to the desired default, so the allocation logic doesn't have to handle it.
+
+## Proposal
+
+The proposal is to implement a new allocator logic that consumes 2 new API Objects: ServiceIPRange
+and ServiceIP, and allow users to dynamically increase the number of Services IPs available by
+creating new ServiceIPRanges.
+
+The new allocator will be able to "automagically" consume IPs from any of this new ranges, we can
+think about this model, as the same as adding more disks to a Storage system to increase the
+capacity.
+
+To simplify the model, make it backwards compatible and to avoid that it can evolve into something
+different and collide with other APIs, like Gateway APIs, we are adding the following constraints:
+
+ - a ServiceIPRange will be immutable after creation
+ - a ServiceIPRange can only be deleted if there are no ServiceIP associated to it
+ - there can not be overlapping IP networks on the ServiceIPRanges created
```

Comment:
my initial idea was implementing 2, that is why I've added an status field to IPrange, so we can indicate the allocators "stop using this range temporary because I'm going to do one operation on it" and we move the complexity to the singleton controller instead of having it per apiserver, avoiding races.

I wanted to start simple, I would love to do 2, but I think that it has to be implemented and documented as:
"When you create a new range that overlaps with an existing range, the controller will take it as a migration operation, 
1. If you have a 10.0.0.0/24 range and you create a new 10.0.0.0/16 range, the controller will understand that you want to scale up the range available, it will verify that the new range 10.0.0.0/16 is ready to be used, and set the status on 10.0.0.0/24 to unready, so the new range can be used"
Same for scale down.
WDYT?

Reply:
so, the constraint will be true, just that the controller will convert it in an update

Reply:
I don't see why the staus lockout is needed - this is not really what status is for.

What does the controller need to do to make it "ready"?  What race are you concerned with?

Each API server has a set of ranges it knows about.  Only non-deleted ranges (no deletionTimestamp) are considered. Ranges might or might not overlap.  The allocator will pick based on a TBD heuristic, but no range in inherently "better" or "worse" than any other.

An allocated service IP implicitly "depends on" all ranges which include it.

The finalizer for a being-deleted range will only be cleared when no IPs depend _exclusively_ on that range.  IOW, after deletion, all IPs are still covered by at least 1 range (unless user action violates that, in which case we can't auto-repair it).  

One race to prevent:
* range X is 10.0.0.0/24
* 3 API servers
* user deletes X via API server 0, blocked by finalizer
* finalizer sees no IPs from X and finishes the deletion
* at the same time, a user creates a service via apiserver 2 
* API server 2 has not yet observed the deletion of X
* API server 2 allocates 10.0.0.42

To block that we either need to wait until all API servers observe the deletion (wall time or a finalizer per apiserver?) or we need to confirm the existence of a covering range on every IP allocation.

Can you think of other sequences of operations that would break?

Reply:
> I don't see why the status lockout is needed - this is not really what status is for.

I was trying to avoid dealing with logic on the allocators to automatically handle overlaps, and put the load on the users, but since we are in full automation mode, I'll remove the status lockout ... we are basically building a routing table and the routes updates come from the API instead of a routing protocol :smile: 

> One race to prevent:

That race is the same scenario as if an user deletes an IPRange with existing IPs allocated, should we handle it differently? is just "bad luck"

> To block that we either need to wait until all API servers observe the deletion (wall time or a finalizer per apiserver?

 ... we can reduce the risk with a wall time, a finalizer per apiserver seems excessive to me, I have doubts that apimachinery folks will like we add that ... we can also document that deleting IPRanges is a disruptive operation and no Services should be created during that operation, with the risk of obtaining an IP that will no longer be valid (if the apiserver takes more than X seconds to see the IPRange delete)

Reply:
> That race is the same scenario as if an user deletes an IPRange with existing IPs allocated, should we handle it differently? is just "bad luck"

If the user overrides the finalizer and breaks the cluster, they earned it.

If we do it occasionally because of "bad luck", we should go to our rooms and think about what we have done.

Reply:
> If we do it occasionally because of "bad luck", we should go to our rooms and think about what we have done.

haha, no, no no, I didn't mean that ... I mean, assuming we do random allocation from each range, is not possible to sync "deleting IPRange" and "create a new Service that allocates new IP", ( I think that two-phase commit or these kind of things solve it but :smile:)  .... so, if we assume that this race is always there, what I'm saying is:

"cluster-admin, bear in mind that when you are deleting an IPRange, there is a time to sync the allocators status, so try to not create Services during that operation, anyway, don't worry, because we'll wait some time to totally remove the range and if you had bad luck we'll warn you, generate an Event informing of the problem and we'll keep the IPRange until the problematic Service is deleted"

We can add an annotation to IPRanges to indicate allocators to not use, this will facilitate operators, the can set the annotation, disable the IPrange and then perform the deletion

Reply:
I think the only simple answer is finalizers waiting, giving time for other apiservers to see it and stop allocating.